File: java/Maybe.java
public abstract class Maybe<A> {
    public boolean isJust;
    public abstract <B> Maybe<B> fmap(Function<A, B> f);
    public abstract <B> Maybe<B> bind(Function<A, Maybe<B>> f);
}

class Nothing<A> extends Maybe<A> {
    public Nothing(){
        isJust = false;
    }

    public <B> Maybe<B> fmap(Function<A, B> f){
        return new Nothing<B>();
    }
    @SuppressWarnings("unchecked")
    public <B> Maybe<B> bind(Function<A, Maybe<B>> f){
        return new Nothing<B>();
    }
}

class Just<A> extends Maybe<A> {
    public final A val;
    public Just(A a){
        val = a;
        isJust = true;
    }

    public <B> Maybe<B> fmap(Function<A, B> f){
        B b = f.apply(val);
        if(b == null) return new Nothing<B>();
        return new Just<B>(b);
    }
    @SuppressWarnings("unchecked")
    public <B> Maybe<B> bind(Function<A, Maybe<B>> f){
        return f.apply(val);
    }
}

File: csharp/Monads/Maybe.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Monads
{
    public interface Maybe<A> 
    {
        Maybe<B> FMap<B>(Func<A, B> f);
        Maybe<B> Bind<B>(Func<A, Maybe<B>> f);
    }

    public class Nothing<A> : Maybe<A>
    {
        public Nothing() { }

        public Maybe<B> FMap<B>(Func<A, B> f)
        {
            return new Nothing<B>();
        }

        public Maybe<B> Bind<B>(Func<A, Maybe<B>> f)
        {
            return new Nothing<B>();
        }
    }

    public class Just<A> : Maybe<A>
    {
        private A a;

        public Just(A a)
        {
            this.a = a;
        }

        public Maybe<B> FMap<B>(Func<A, B> f)
        {
            try
            {
                var result = f.Invoke(a);
                return new Just<B>(result);
            }
            catch (Exception)
            {
                return new Nothing<B>();
            }
        }

        public Maybe<B> Bind<B>(Func<A, Maybe<B>> f)
        {
			try 
			{
				var result = f.Invoke(a);
				return result;
			}
			catch (Exception)
			{
				return new Nothing<B> ();
			}
        }

        public A Value { get { return a; } }
    }
}
